### ISSUE TITLE
How to handle composite keys, custom ID

### ISSUE THREAD

**Issue by oojacoboo** [thecodingmachine/graphqlite#223]
So, I'm faced with an issue I'm really not sure how best to approach.  I started playing around with a custom type mapper as well as the field middleware.  I'm able to target the field in question, but I'm not really sure how it'd be best to handle this situation.

Basically, I'd like to have the `ID` be a composite key represented as a UUID.  I thought maybe I could handle this as a scalar, but I don't want to alter how my domain objects are written to return an object.  Basically, I want my, `getId` in my domain object to continue to return an `int`, but I want GraphQL to return a UUID, which is a composite key.  GraphQL would also receive the composite key as an `ID` per usual, just with some of my custom logic attached, simply by decoding the UUID and pushing the state into the `ID` object.

Basically, I'd like to take control over how the `ID` is handled.  But, it seems that custom type mappers are last in line, instead of first - how they should be in my opinion.

Any ideas on how I can handle this?  Can I just extend the `ID`?  Will that work for input and output?  Or, do I need to go down the type mapper or field middleware routes?

**Comment by mailopl** [comment]
Let me go little of topic here - any reason why you're not using ValueObject's in your domain entity so you could map your CompositeIdentifier (instead of string) to GQL.ID as you mentioned?

**Comment by oojacoboo** [comment]
@mailopl I looked into this, but we don't have access to the data necessary to create these value objects with our ORM, during hydration of the objects.  So, unfortunately it doesn't seem like that's really an option for us.  We looked into it and even tried to implement it as such.

We ended up adding a new method on the domain entity for the composite ID and exposed it as the `ID`.  It ended up being an incredibly simple solution and probably for the best since we're still able to maintain our auto-increment id on the entity as well.
